src/protocols/bdx/BDXMessageUtils.cpp

```diff
@@ -0,0 +1,444 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      Implements utility methods for working with some complex BDX messages.
+ */
+
+#include <protocols/bdx/BDXMessageUtils.h>
+
+#include <support/BufBound.h>
+#include <support/BufferReader.h>
+#include <support/CodeUtils.h>
+
+#define VERSION_MASK 0x0F
+#define SENDER_DRIVE_MASK 0x10
+#define RECEIVER_DRIVE_MASK 0x20
+#define ASYNC_MASK 0x40
+#define CONTROL_MODE_MASK 0xF0
+
+#define DEFLEN_MASK 0x01
+#define START_OFFSET_MASK 0x02
+#define WIDERANGE_MASK 0x10
+
+using namespace chip;
+using namespace chip::BDX;
+using namespace chip::Encoding::LittleEndian;
+
+CHIP_ERROR TransferInit::Pack(System::PacketBuffer * aBuffer)
+{
+    CHIP_ERROR err   = CHIP_NO_ERROR;
+    uint8_t rangeCtl = 0, ptcByte = 0;
+
+    // BufBound has to be initialized with aBuffer so VerifyOrExit can't be used here.
+    if (aBuffer == NULL)
```

Comment:
instead of this, how about having the function take a reference instead?

Reply:
I think using a pointer was an old Weave convention, but I don't see why we couldn't use a reference here

Reply:
Ah, Weave convention was to pass a pointer because the only way to obtain PacketBuffers are the various New() methods which return a pointer. So is it better to force the caller to dereference the pointer before calling the function, or do the NULL check in the method?

Reply:
the former, since presumably the code allocating/creating/receiving the packetbuffer has already checked for allocation failure?

Reply:
Yea that's a good point. I don't think we expect for this function to be called with a NULL packetbuffer